include/yaclib/task.hpp

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <yaclib/container/intrusive_node.hpp>
+
+namespace yaclib {
+
+class ITask : private container::detail::IntrusiveNode<ITask> {
+ public:
+  virtual void Call() = 0;
+
+  virtual ~ITask() = default;
+
+  container::detail::IntrusiveNode<ITask>* AsNode() {
+    return static_cast<container::detail::IntrusiveNode<ITask>*>(this);
+  }
+};
+
+using ITaskPtr = std::shared_ptr<ITask>;
+
+namespace detail {
+
+template <typename Functor>
+class Task final : public ITask {
+ public:
+  Task(Functor&& f) : _f{std::move(f)} {
+  }
+
+  Task(const Functor& f) : _f{f} {
+  }
+
+  void Call() final {
+    return _f();
+  }
+
+ private:
+  Functor _f;
+};
+
+}  // namespace detail
+
+template <typename Functor>
+ITaskPtr CreateTask(Functor&& f) {
+  using FunctorT = std::remove_reference_t<Functor>;
+  return std::make_shared<detail::Task<FunctorT>>(std::forward<Functor>(f));
+}
+
+template <typename Class>
+ITaskPtr CreateTask(Class& instance, void (Class::*f)()) {
+  auto instance_ptr = &instance;
+  return CreateTask([instance_ptr, f]() mutable {
+    return (instance_ptr->*f)();
+  });
+}
+
+template <typename Class>
+ITaskPtr CreateTask(const Class& instance, void (Class::*f)() const) {
```

Comment:
Question
Is this overload really necessary?

Reply:
Yes, other way there won't be any matching function for call to `CreateTask` from const method: these rows https://github.com/YACLib/YACLib/blob/536fd666772c27907226ba18b67bc245c60d87a2/test/task.cpp#L85 https://github.com/YACLib/YACLib/blob/536fd666772c27907226ba18b67bc245c60d87a2/test/task.cpp#L90 will cause compile errors.